src/Testable/Navigation.elm

```diff
@@ -0,0 +1,156 @@
+module Testable.Navigation exposing (..)
+
+import List.Extra
+import
+    -- This "unused" import is required for the native bindings
+    Native.Testable.Navigation
+import Navigation exposing (Location)
+import Regex exposing (HowMany(..), find, regex)
+
+
+type alias History =
+    ( Int, List Location )
+
+
+type Msg
+    = Jump Int
+    | New String
+    | Modify String
+    | Visit String
+
+
+type ReturnMsg
+    = NoOp
+    | ReturnLocation Location
+    | TriggerLocationMsg Location
+
+
+currentLocation : History -> Location
+currentLocation ( index, history ) =
+    List.Extra.getAt index history
+        |> Maybe.withDefault (getLocation "")
+
+
+update : Msg -> History -> ( History, ReturnMsg )
+update msg ( index, history ) =
+    let
+        location =
+            currentLocation ( index, history )
+    in
+    case msg of
+        Jump n ->
+            let
+                nextIndex =
+                    min (max (index + n) 0) (List.length history - 1)
+
+                nextLocation =
+                    List.Extra.getAt nextIndex history
+                        |> Maybe.withDefault (getLocation "")
+            in
+            ( ( nextIndex, history ), TriggerLocationMsg nextLocation )
+
+        New url ->
+            let
+                nextLocation =
+                    setLocation url location
+
+                modifiedHistory =
+                    List.take (index + 1) history ++ [ nextLocation ]
+            in
+            ( ( index + 1, modifiedHistory ), ReturnLocation nextLocation )
+
+        Modify url ->
+            let
+                nextLocation =
+                    setLocation url location
+
+                modifiedHistory =
+                    history
+                        |> List.Extra.updateIfIndex ((==) index) (always nextLocation)
+
+                effect =
+                    if nextLocation == location then
+                        NoOp
+                    else
+                        ReturnLocation nextLocation
+            in
+            ( ( index, modifiedHistory ), effect )
+
+        Visit url ->
+            let
+                ( modifiedHistory, _ ) =
+                    update (New url) ( index, history )
+            in
+            ( modifiedHistory, TriggerLocationMsg (currentLocation modifiedHistory) )
+
+
+init : History
+init =
+    ( 0, [ initialLocation ] )
+
+
+initialLocation : Location
+initialLocation =
+    getLocation "https://elm.testable/"
```

Comment:
yeah, so, one problem is that both `Html.program` and `Navigation.program` return a `Program Never Model Msg`, I don't know a way to tell them apart.

also, I believe you could use things like `Navigation.back` on a regular `Html.program`, the difference is that by using `Navigation.program` you can listen to browser url change events.

the thing is: an application always has a url, even a non-navigational one, so what should we do? maybe add a `startWithLocation` function but fallback to the default one if not provided?